src/plugins/intel_cpu/src/nodes/rope.cpp

```diff
@@ -7,17 +7,281 @@
 #include "shape_inference/shape_inference_internal_dyn.hpp"
 #include "utils/plain_tensor.hpp"
 
 #include <chrono>
 #include <string>
 #include <vector>
 
+using namespace dnnl;
+using namespace dnnl::impl;
+using namespace dnnl::impl::cpu::x64;
+using namespace dnnl::impl::utils;
+using namespace Xbyak;
+
 namespace ov {
 namespace intel_cpu {
 namespace node {
 
+struct jit_rotary_compile_params {
+    ov::element::Type src_prc;
+    ov::element::Type dst_prc;
+    size_t rotary_ndims;
+    bool interleave;
+    bool mix_cos_sin;
+};
+struct jit_rotary_call_args {
+    void* src;
+    float* cos;
+    float* sin;
+    void* dst;
+};
+struct jit_uni_rotary_kernel {
+    void (*ker_)(const jit_rotary_call_args*);
+    void operator()(const jit_rotary_call_args* call_args) {
+        assert(ker_);
+        ker_(call_args);
+    }
+    explicit jit_uni_rotary_kernel(const jit_rotary_compile_params& jcp) : ker_(nullptr), jcp_(jcp) {}
+    virtual ~jit_uni_rotary_kernel() {}
+    virtual void create_ker() = 0;
+    jit_rotary_compile_params jcp_;
+};
+
+#ifdef OPENVINO_ARCH_X86_64
+
+template <cpu_isa_t isa>
+struct jit_rotary_kernel : public jit_uni_rotary_kernel, public jit_generator {
+    DECLARE_CPU_JIT_AUX_FUNCTIONS(jit_rotary_kernel)
+    static constexpr size_t vec_size = dnnl::impl::cpu::x64::cpu_isa_traits<isa>::vlen / sizeof(float);
+    explicit jit_rotary_kernel(const jit_rotary_compile_params& jcp) : jit_uni_rotary_kernel(jcp), jit_generator(jit_name()) {}
+    virtual ~jit_rotary_kernel() {}
+    void create_ker() override {
+        jit_generator::create_kernel();
+        ker_ = (decltype(ker_))jit_ker();
+    }
+
+private:
+    using Vmm = typename dnnl::impl::utils::conditional3<isa == cpu_isa_t::sse41, Xmm, isa == cpu_isa_t::avx2, Ymm, Zmm>::type;
+    void generate() override {
+        this->preamble();
+#define GET_OFF(field) offsetof(jit_rotary_call_args, field)
+        mov(reg_src, ptr[reg_params + GET_OFF(src)]);
+        mov(reg_cos, ptr[reg_params + GET_OFF(cos)]);
+        mov(reg_sin, ptr[reg_params + GET_OFF(sin)]);
+        mov(reg_dst, ptr[reg_params + GET_OFF(dst)]);
+        uni_vpxor(vmm_src0, vmm_src0, vmm_src0);
+        uni_vpxor(vmm_src1, vmm_src1, vmm_src1);
+        uni_vpxor(vmm_cos, vmm_cos, vmm_cos);
+        uni_vpxor(vmm_sin, vmm_sin, vmm_sin);
+        if (jcp_.interleave) {
+            // dst: 0-2 4-6 8-10 12-14 16-18 20-22 24-26 28-30 ->
+            // lower 64bit/128 lane
+            //      0-2        4-6        8-10       12-14
+            // higher 64bit/128 lane
+            //           16-18      20-22      24-26       28-30
+            static const uint64_t mask_zmm[] = {
+                0, 4, 1, 5, 2, 6, 3, 7
+            };
+            if (isa == cpu_isa_t::avx512_core) {
+                mov(reg_tmp, reinterpret_cast<uintptr_t>(mask_zmm));
+                uni_vmovups(vmm_idx, ptr[reg_tmp]);
+            }
+            auto half_rotary_ndims = jcp_.rotary_ndims / 2;
```

Comment:
I am wondering what is the typical value for `rotary_ndims`?

Reply:
The vaule of some models:
llama2-7b/13b/mistral-7b: 128
chatglm2/3-6b: 64
qwen-7b: 128
gptj-6b: 64
gemm-7b: 192
gptneox-20b: 24